templates/cmake.mpd

```diff
@@ -0,0 +1,158 @@
+cmake_minimum_required(VERSION <%cmake_minimum_required(3.12.0)%>)
+
+<%marker(top)%>
+project(<%project_name%> <%language%>)
+option(BUILD_SHARED_LIBS "Build using shared libraries" <%if(need_staticflags)%>OFF<%else%>ON<%endif%>)
+
+<%if(pre_find_package)%>
+<%pre_find_package%>
+<%endif%>
+<%foreach(packages)%>
+find_package(<%package%> REQUIRED)
+<%endfor%>
+
+set(SOURCE_FILES <%source_files%>)
+<%if(libs || lit_libs || pure_libs)%>
+set(TARGET_LINK_LIBRARIES<%if(libs)%> <%foreach(configurations)%><%fornotfirst("                          ")%>$<$<CONFIG:<%configuration%>>:<%foreach(libs)%><%fornotfirst(" ")%><%lib%>${LIBRARY_DECORATOR}<%lib_modifier%><%endfor%>><%fornotlast("\n")%><%endfor%><%endif%><%if(lit_libs)%> <%lit_libs%><%endif%><%if(pure_libs)%> <%pure_libs%><%endif%>)
+<%endif%>
+set(PROJECT_TARGET <%if(exename)%><%exename%><%else%><%if(sharedname)%><%sharedname%>${LIBRARY_DECORATOR}<%else%><%if(staticname)%><%staticname%>${LIBRARY_DECORATOR}<%else%><%project_name%>${LIBRARY_DECORATOR}<%endif%><%endif%><%endif%>)
+<%marker(macros)%>
+
+<%if(exeout)%>
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY <%env_exeout%>)
+<%if(use_lib_modifier)%>
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG <%env_exeout%>)
+<%endif%>
+<%else%>
+<%if(libout)%>
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY <%env_libout%>)
+<%if(use_lib_modifier)%>
+<%foreach(configurations)%>
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_<%uc(configuration)%> <%env_libout%>)
+<%endfor%>
+<%endif%>
+<%endif%>
+<%if(dllout)%>
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY <%env_dllout%>)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY <%env_dllout%>)
+<%if(use_lib_modifier)%>
+<%foreach(configurations)%>
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_<%uc(configuration)%> <%env_dllout%>)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_<%uc(configuration)%> <%env_dllout%>)
+<%endfor%>
+<%endif%>
+<%else%>
+<%if(libout)%>
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY <%env_libout%>)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY <%env_libout%>)
+<%if(use_lib_modifier)%>
+<%foreach(configurations)%>
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_<%uc(configuration)%> <%env_libout%>)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_<%uc(configuration)%> <%env_libout%>)
+<%endfor%>
+<%endif%>
+<%endif%>
+<%endif%>
+<%endif%>
```

Comment:
If they aren't set in MPC, then they wouldn't be emitted.  Right?

Reply:
They would be put somewhere in [`CMAKE_CURRENT_BINARY_DIR`](https://cmake.org/cmake/help/latest/variable/CMAKE_CURRENT_BINARY_DIR.html). This probably requires more changes because we're manually dealing with paths. So I don't know how much we'd want to do now, but even basic support for out of source builds are something I'd love to see.

Reply:
To be clear, the `CMAKE_ARCHIVE_OUTPUT_DIRECTORY`, `CMAKE_LIBRARY_OUTPUT_DIRECTORY`, and `CMAKE_RUNTIME_OUTPUT_DIRECTORY` settings will only be emitted to the CMakeLists.txt if the user set `libout` or `dllout` in the .mpc file.  `CMAKE_RUNTIME_OUTPUT_DIRECTORY` will be emitted if the user set 'exeout'.

If none of those values are set, then none of those settings will be emitted to the CMakeLists.txt.  They could even be overridden on the MPC command line when generating.  i.e., `$ACE_ROOT/bin/mwc.pl -type cmake -value_project libout=`

Reply:
I have a project that looks like this:

```
project(test_lib) {
  Header_Files {
    test_lib.h
  }

  Source_Files {
    test_lib.cpp
  }
}

project(test) {
  libs += test_lib
  after += test_lib

  Source_Files {
    test.cpp
  }
}

```

It's putting this in the CMakeLists files:

```
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_DEBUG ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_RELEASE ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_DEBUG ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE ${CMAKE_CURRENT_SOURCE_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELEASE ${CMAKE_CURRENT_SOURCE_DIR})
```

And putting both the library and the executable in the source directory.

Reply:
Right.  Because the default value for `libout` is the current directory (see https://github.com/DOCGroup/MPC/blob/master/config/global.mpb).  In my previous comment, I addressed this directly.  If you generate the CMake files using `$ACE_ROOT/bin/mwc.pl -type cmake -value_project libout=` those lines will not be there.